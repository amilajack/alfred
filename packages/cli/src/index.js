// @flow
import path from 'path';
import fs from 'fs';
import { formatPkgJson } from '@alfred/core/lib/config';
import handlebars from 'handlebars';
import type { InterfaceState } from '@alfred/core';

const TEMPLATES_DIR = path.resolve(__dirname, '../templates');

export * from './helpers/parse-input';

async function compileTemplate(templateFilename: string) {
  const source = await fs.promises.readFile(
    path.resolve(TEMPLATES_DIR, templateFilename)
  );
  return handlebars.compile(source.toString(), { noEscape: true });
}

export async function addEntrypoints(
  rawTemplateData: Object,
  root: string,
  entrypointInterfaceStates: Array<InterfaceState>
) {
  const [
    APP_TEMPLATE,
    LIB_TEMPLATE,
    APP_BROWSER_HTML_TEMPLATE,
    TEST_TEMPLATE
  ] = await Promise.all(
    ['app.js.hbs', 'lib.js.hbs', 'index.html.hbs', 'test.js.hbs'].map(
      compileTemplate
    )
  );

  await Promise.all(
    entrypointInterfaceStates.map(interfaceState => {
      const { projectType, target } = interfaceState;
      const isApp = projectType === 'app';
      const isBrowser = target === 'browser';

      const templateData = {
        ...rawTemplateData,
        project: { projectType, target, isApp, isBrowser }
      };

      const writeIndexHtml = () => {
        if (isApp && isBrowser) {
          const content = APP_BROWSER_HTML_TEMPLATE(templateData);
          return fs.promises.writeFile(
            path.join(root, './src/index.html'),
            content
          );
        }
        return Promise.resolve();
      };

      return Promise.all(
        [
          {
            file: `./src/${projectType}.${target}.js`,
            content: (isApp ? APP_TEMPLATE : LIB_TEMPLATE)(templateData)
          },
          {
            file: `./tests/${projectType}.${target}.spec.js`,
            content: TEST_TEMPLATE(templateData)
          }
        ]
          .map(({ file, content }) =>
            fs.promises.writeFile(path.join(root, file), content)
          )
          .concat([writeIndexHtml()])
      );
    })
  );
}

// This function lives here because it is used in both tests and the cli
export async function addBoilerplate(templateData: Object, root: string) {
  const [
    GITIGNORE_TEMPLATE,
    NPM_TEMPLATE,
    README_TEMPLATE,
    EDITORCONFIG_TEMPLATE
  ] = await Promise.all(
    [
      '.gitignore.hbs',
      'package.json.hbs',
      'README.md.hbs',
      '.editorconfig.hbs'
    ].map(compileTemplate)
  );

  await Promise.all(
    [
      {
        file: '.gitignore',
        content: GITIGNORE_TEMPLATE(templateData)
      },
      {
        file: '.editorconfig',
        content: EDITORCONFIG_TEMPLATE(templateData)
      },
      {
        file: 'README.md',
        content: README_TEMPLATE(templateData)
      }
    ].map(({ file, content }) =>
      fs.promises.writeFile(path.join(root, file), content)
    )
  );

  const pkgPath = path.join(root, 'package.json');
  const formattedPkg = await formatPkgJson(
    JSON.parse(NPM_TEMPLATE(templateData))
  );
  await fs.promises.writeFile(pkgPath, formattedPkg);
  const { projectType, target } = templateData.project;

  await addEntrypoints(templateData, root, [
    { projectType, target, env: 'production' }
  ]);
}

export default function getSingleSubcommandFromArgs(
  args: Array<string>
): string {
  switch (args.length) {
    case 0: {
      throw new Error('One subcommand must be passed');
    }
    case 1: {
      break;
    }
    default: {
      throw new Error('Only one subcommand can be passed');
    }
  }

  return args[0];
}
