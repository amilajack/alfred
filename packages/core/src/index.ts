import Project from './project';
import { ProjectInterface } from '@alfred/types';

if (typeof jest !== 'undefined') {
  require('source-map-support/register');
}

export default function alfred(projectDir?: string): Promise<ProjectInterface> {
  const project = new Project(projectDir);
  return project.init();
}

// @TODO @REFACTOR Move this to another module
export { formatPkgJson } from './project';
