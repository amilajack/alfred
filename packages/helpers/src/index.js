// @flow
/* eslint import/prefer-default-export: off */
/* eslint import/no-dynamic-require: off */
import path from 'path';
import opn from 'opn';
import pkgUp from 'pkg-up';
import childProcess from 'child_process';
import type {
  configFileType,
  configType,
  CtfMap
} from '@alfred/core/lib/types';

/**
 * Get the root of a project from the current working directory
 */
export function searchProjectRoot(startingSearchDir: string = process.cwd()) {
  const pkgPath = pkgUp.sync({
    cwd: startingSearchDir
  });
  if (!pkgPath) {
    throw new Error(
      `Alfred project root could not be found from "${startingSearchDir}".

      Make sure you are inside an Alfred project.`
    );
  }
  return path.dirname(pkgPath);
}

export function getConfigByConfigName(
  configName: string,
  configFiles: Array<configFileType>
) {
  const config = configFiles.find(e => e.name === configName);
  if (!config) throw new Error(`Cannot find config by name "${configName}"`);
  return config;
}

export function getConfigPathByConfigName(
  configName: string,
  configFiles: Array<configFileType>
) {
  const config = configFiles.find(e => e.name === configName);
  if (!config) throw new Error(`Cannot find config by name "${configName}"`);
  return config.path;
}

/**
 * Map the environment name to a short name, which is one of ['dev', 'prod', 'test']
 * @TODO: Should be moved to CLI
 */
export function mapEnvToShortName(envName: string): string {
  switch (envName) {
    case 'production': {
      return 'prod';
    }
    case 'development': {
      return 'dev';
    }
    case 'test': {
      return 'test';
    }
    default: {
      throw new Error(`Unsupported environment "${envName}"`);
    }
  }
}

export function mapShortNameEnvToLongName(envName: string): string {
  switch (envName) {
    case 'prod': {
      return 'production';
    }
    case 'dev': {
      return 'development';
    }
    default: {
      throw new Error(`Unsupported short name environment "${envName}"`);
    }
  }
}

/*
 * Intended to be used for testing purposes
 */
export function getConfigs(ctf: CtfMap): Array<configType> {
  return Array.from(ctf.values())
    .map(ctfNode => ctfNode.configFiles || [])
    .reduce((p, c) => [...p, ...c], [])
    .map(e => e.config);
}

export function getConfigsBasePath(projectRoot: string): string {
  return path.join(projectRoot, '.configs');
}

export function requireConfig(configName: string): any {
  try {
    // $FlowFixMe
    return require(`alfred-config-${configName}`);
  } catch (e) {
    try {
      // $FlowFixMe
      return require(configName);
    } catch (_e) {
      throw new Error(
        `Could not resolve "${configName}" module or "eslint-config-${configName}" module`
      );
    }
  }
}

/**
 * Get the name of the package JSON
 * @param {string} pkgName The name of the package
 * @param {string} binName The property of the bin object that we want
 */
export async function getPkgBinPath(pkgName: string, binName: string) {
  const pkgPath = require.resolve(pkgName);
  const pkgJsonPath = await pkgUp({ cwd: pkgPath });

  const { bin } = require(pkgJsonPath);
  if (!bin) {
    throw new Error(
      `Module "${pkgName}" does not have a binary because it does not have a "bin" property in it's package.json`
    );
  }

  return path.join(
    path.dirname(pkgJsonPath),
    typeof bin === 'string' ? bin : bin[binName]
  );
}

export function execCommand(cmd: string) {
  return childProcess.execSync(cmd, { stdio: 'inherit' });
}

export async function openInBrowser(url: string, browser: any) {
  // Don't open new tab when running end to end tests. This prevents hundreds
  // of tabs from being opened.
  if (process.env.E2E_CLI_TEST) return;

  try {
    const options = typeof browser === 'string' ? { app: browser } : undefined;
    await opn(url, options);
  } catch (err) {
    console.error(`Unexpected error while opening in browser: ${browser}`);
    console.error(err);
  }
}
