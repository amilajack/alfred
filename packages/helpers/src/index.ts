/* eslint import/prefer-default-export: off, import/no-dynamic-require: off */
import path from 'path';
import open from 'open';
import childProcess from 'child_process';
import serialize from 'serialize-javascript';
import {
  AlfredConfigWithUnresolvedInterfaces,
  ProjectInterface,
  ConfigInterface,
  PkgWithAllDeps,
  DependencyType,
  DependencyTypeFull,
  Dependencies
} from '@alfred/types';

export class EnhancedMap<K, V> extends Map<K, V> {
  map(fn: (item: V, idx: number, items: [K, V][]) => V): EnhancedMap<K, V> {
    const newMap = new EnhancedMap<K, V>();
    Array.from(this.entries()).forEach(([key, val], idx, items) => {
      newMap.set(key, fn(val, idx, items));
    });
    return newMap;
  }
}

export function configSerialize(config: string | Record<string, any>): string {
  return serialize(config, { unsafe: true });
}

export function configStringify(configStr: TemplateStringsArray): string {
  return ['[alfred]', configStr, '[alfred]']
    .join('')
    .trim()
    .replace(/\n/g, ' ');
}

export function configToEvalString(serializedConfig: string): string {
  return serializedConfig
    .replace(/"\[alfred\]/g, '')
    .replace(/\[alfred\]"/g, '');
}

export function fromPkgTypeToFull(
  fromPkgType: DependencyType
): DependencyTypeFull {
  switch (fromPkgType) {
    case 'peer':
      return 'peerDependencies';
    case 'dev':
      return 'devDependencies';
    case 'dep':
      return 'dependencies';
    default: {
      throw new Error('Pkg type must be one of peer, dev, or dep');
    }
  }
}

/**
 * Map the environment name to a short name, which is one of ['dev', 'prod', 'test']
 */
export function mapEnvToShortName(envName: string): string {
  switch (envName) {
    case 'production': {
      return 'prod';
    }
    case 'development': {
      return 'dev';
    }
    case 'test': {
      return 'test';
    }
    default: {
      throw new Error(`Unsupported environment "${envName}"`);
    }
  }
}

export function mapShortNameEnvToLongName(envName: string): string {
  switch (envName) {
    case 'prod': {
      return 'production';
    }
    case 'dev': {
      return 'development';
    }
    default: {
      throw new Error(`Unsupported short name environment "${envName}"`);
    }
  }
}

export function getConfigsBasePath(
  project: ProjectInterface,
  config: ConfigInterface = project.config
): string {
  return path.join(project.root, config.configsDir);
}

export function requireConfig(
  configName: string
): AlfredConfigWithUnresolvedInterfaces {
  try {
    return require(`alfred-config-${configName}`);
  } catch (e) {
    try {
      return require(configName);
    } catch (_e) {
      throw new Error(
        `Could not resolve "${configName}" module or "eslint-config-${configName}" module`
      );
    }
  }
}

export function execCmdInProject(
  project: ProjectInterface,
  cmd: string,
  opts: Record<string, any> = {}
): Buffer {
  return childProcess.execSync(cmd, {
    stdio: 'inherit',
    cwd: project.root,
    ...opts
  });
}

/**
 * Get the name of the package JSON
 * @param pkgName - The name of the package
 * @param binName - The property of the bin object that we want
 */
export async function getPkgBinPath(
  project: ProjectInterface,
  pkgName: string
): Promise<string> {
  switch (project.config.npmClient) {
    case 'npm': {
      return execCmdInProject(project, `npm bin ${pkgName}`, { stdio: 'pipe' })
        .toString()
        .trim();
    }
    case 'yarn': {
      return execCmdInProject(project, `yarn bin ${pkgName}`, { stdio: 'pipe' })
        .toString()
        .trim();
    }
    default: {
      throw new Error(`Unsupported npmClient "${project.config.npmClient}"`);
    }
  }
}

export async function openUrlInBrowser(
  url: string,
  browser: string
): Promise<void> {
  // Don't open new tab when running end to end tests. This prevents hundreds
  // of tabs from being opened.
  if (process.env.ALFRED_E2E_CLI_TEST) return;

  try {
    const options = typeof browser === 'string' ? { app: browser } : undefined;
    await open(url, options);
  } catch (err) {
    console.error(`Unexpected error while opening in browser: ${browser}`);
    console.error(err);
  }
}

export function getDepsFromPkg(
  pkgs: string | string[],
  pkg: PkgWithAllDeps,
  fromPkgType: DependencyType = 'dev'
): Dependencies {
  const normalizedPkgNames = Array.isArray(pkgs) ? pkgs : [pkgs];
  const fromPkgTypeFull = fromPkgTypeToFull(fromPkgType);

  if (!(fromPkgTypeFull in pkg)) {
    throw new Error(`Given package.json does not have ${fromPkgTypeFull}`);
  }

  normalizedPkgNames.forEach(pkgName => {
    if (!(pkgName in pkg[fromPkgTypeFull])) {
      throw new Error(
        `Package "${pkgName}" does not exist in ${fromPkgTypeFull} of skill package.json`
      );
    }
  });

  return Object.fromEntries(
    normalizedPkgNames.map(pkgName => [pkgName, pkg[fromPkgTypeFull][pkgName]])
  );
}

export function serialPromises(fns: Array<() => Promise<any>>): Promise<any> {
  return fns.reduce(
    (promise: Promise<any>, fn) =>
      // eslint-disable-next-line promise/no-nesting
      promise.then(result => fn().then(Array.prototype.concat.bind(result))),
    Promise.resolve([])
  );
}
