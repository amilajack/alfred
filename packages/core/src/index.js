import path from 'path';
import rimraf from 'rimraf';
import fs from 'fs';
import childProcess from 'child_process';
import { getProjectRoot } from '@alfredpkg/cli/lib/helpers/CLI';
import jestCtf from '@alfredpkg/skill-jest';
import babel from '@alfredpkg/skill-babel';
import webpack from '@alfredpkg/skill-webpack';
import eslint from '@alfredpkg/skill-eslint';
import react from '@alfredpkg/skill-react';
import prettier from '@alfredpkg/skill-prettier';
import rollup from '@alfredpkg/skill-rollup';
import lodashCtf from '@alfredpkg/skill-lodash';
import mergeConfigs from '@alfredpkg/merge-configs';
import pkgUp from 'pkg-up';
import lodash from 'lodash';
import type { AlfredConfig } from '@alfredpkg/cli';

// All the possible interface states
export const INTERFACE_STATES = [
  {
    projectType: 'app',
    target: 'browser',
    env: 'production'
  },
  {
    projectType: 'app',
    target: 'lib',
    env: 'production'
  },
  {
    projectType: 'lib',
    target: 'browser',
    env: 'production'
  },
  {
    projectType: 'app',
    target: 'node',
    env: 'production'
  },
  {
    projectType: 'app',
    target: 'browser',
    env: 'development'
  },
  {
    projectType: 'app',
    target: 'lib',
    env: 'development'
  },
  {
    projectType: 'lib',
    target: 'browser',
    env: 'development'
  },
  {
    projectType: 'app',
    target: 'node',
    env: 'development'
  }
];

export const CORE_CTFS = {
  babel,
  webpack,
  eslint,
  prettier,
  jest: jestCtf,
  react,
  rollup,
  lodash: lodashCtf
};

// @TODO send the information to a crash reporting service (like sentry.io)
process.on('unhandledRejection', err => {
  throw err;
});

export type InterfaceState = {
  // Flag name and argument types
  env: 'production' | 'development' | 'test',
  // All the supported targets a `build` skill should build
  target: 'browser' | 'node' | 'electron' | 'react-native',
  // Project type
  projectType: 'lib' | 'app'
};

export type InterfaceInputType = Array<
  string | [string, { [x: string]: string }]
>;
export type NormalizedInterfacesType = Array<{
  name: string,
  module: Object
}>;

export function normalizeInterfacesOfSkill(
  interfaces: InterfaceState
): NormalizedInterfacesType {
  if (!interfaces) return [];
  // `interfaces` is an array
  if (Array.isArray(interfaces)) {
    // @HACK Check if the array is alread formatted with this function by
    //       checking if name property exists
    if (interfaces[0] && interfaces[0].name) {
      return interfaces;
    }
    return interfaces.map(e => {
      if (typeof e === 'string') {
        return {
          name: e,
          module: require(e), // eslint-disable-line
        };
      }
      if (Array.isArray(e)) {
        if (e.length !== 2) {
          throw new Error(
            'Interface tuple config must have exactly two elements'
          );
        }
        const [name, config] = e;
        return {
          name,
          module: require(name), // eslint-disable-line
          config
        };
      }
      throw new Error('Interface config must be either an array or a string');
    });
  }
  throw new Error(
    `".interfaces" property must be an array of strings or an array of arrays. Received "${interfaces}"`
  );
}

export type configType =
  | string
  | {
      [x: string]: any
    };

export type configFileType = {
  // The "friendly name" of a file. This is the name that
  // other CTFs will refer to config file by.
  name: string,
  // The relative path of the file the config should be written to
  path: string,
  // The value of the config
  config: configType,
  // The type of the config file. Defaults to 'json'
  fileType: 'module' | 'string' | 'json',
  // Allow the config to be written to user's `./configs` directory
  write: boolean
};

type UsingInterface = {|
  interfaces: InterfaceInputType,
  subcommand: string,
  hooks: {
    call: ({
      fileConfigPath: string,
      config: configFileType,
      alfredConfig: Object,
      interfaceState: InterfaceState,
      subcommand: string
    }) => string,
    install?: () => void
  }
|};

// @flow
type RequiredCtfNodeParams = {|
  name: string,
  dependencies: {
    [x: string]: any
  },
  devDependencies: {
    [x: string]: any
  },
  description: string,
  configFiles: Array<configFileType>,
  ctfs: {
    [x: string]: (
      RequiredCtfNodeParams,
      Map<string, RequiredCtfNodeParams>
    ) => RequiredCtfNodeParams
  }
|};

export type CtfNode =
  | RequiredCtfNodeParams
  | {| ...UsingInterface, ...RequiredCtfNodeParams |};

export type CtfMap = Map<string, CtfNode>;

export function getConfigByConfigName(
  configName: string,
  configFiles: Array<configFileType>
) {
  const config = configFiles.find(e => e.name === configName);
  if (!config) throw new Error(`Cannot find config by name "${configName}"`);
  return config;
}

/**
 * Map the environment name to a short name, which is one of ['dev', 'prod', 'test']
 * @TODO: Should be moved to CLI
 */

export function mapEnvToShortName(envName: string): string {
  switch (envName) {
    case 'production': {
      return 'prod';
    }
    case 'development': {
      return 'dev';
    }
    case 'test': {
      return 'test';
    }
    default: {
      throw new Error(`Unsupported environment "${envName}"`);
    }
  }
}

export function mapShortNameEnvToLongName(envName: string): string {
  switch (envName) {
    case 'prod': {
      return 'production';
    }
    case 'dev': {
      return 'development';
    }
    default: {
      throw new Error(`Unsupported short name environment "${envName}"`);
    }
  }
}

/**
 * Get the name of the package JSON
 * @param {string} pkgName The name of the package
 * @param {string} binName The property of the bin object that we want
 */
export async function getPkgBinPath(pkgName: string, binName: string) {
  const pkgPath = require.resolve(pkgName);
  const pkgJsonPath = await pkgUp(pkgPath);

  const { bin } = require(pkgJsonPath); // eslint-disable-line
  if (!bin) {
    throw new Error(
      `Module "${pkgName}" does not have a binary because it does not have a "bin" property in it's package.json`
    );
  }

  return path.join(
    path.dirname(pkgJsonPath),
    typeof bin === 'string' ? bin : bin[binName]
  );
}

const configsBasePath = path.join(getProjectRoot(), '.configs');

export function getConfigPathByConfigName(
  configName: string,
  configFiles: Array<configFileType>
) {
  const config = configFiles.find(e => e.name === configName);
  if (!config) throw new Error(`Cannot find config by name "${configName}"`);
  return config.path;
}

type CtfHelpers = {
  findConfig: (configName: string) => configFileType,
  addDependencies: ({ [x: string]: string }) => { [x: string]: string },
  addDevDependencies: ({ [x: string]: string }) => { [x: string]: string },
  extendConfig: (x: string) => CtfNode,
  replaceConfig: (x: string, configReplacement: configType) => CtfNode
};

export const AddCtfHelpers: CtfHelpers = {
  findConfig(configName: string) {
    const config = this.configFiles.find(
      configFile => configFile.name === configName
    );
    if (!config) {
      throw new Error(`Cannot find config with name "${configName}"`);
    }
    return config;
  },
  extendConfig(
    configName: string,
    configExtension: { [x: string]: string } = {}
  ): CtfNode {
    const foundConfig = this.findConfig(configName);
    const mergedConfigFile = mergeConfigs({}, foundConfig, {
      config: configExtension
    });
    const configFiles = this.configFiles.map(configFile =>
      configFile.name === configName ? mergedConfigFile : configFile
    );
    return lodash.merge({}, this, {
      configFiles
    });
  },
  replaceConfig(configName: string, configReplacement: configType) {
    const configFiles = this.configFiles.map(configFile =>
      configFile.name === configName ? configReplacement : configFile
    );
    return {
      ...this,
      configFiles
    };
  },
  addDependencies(dependencies) {
    return lodash.merge({}, this, {
      dependencies
    });
  },
  addDevDependencies(devDependencies) {
    return lodash.merge({}, this, {
      devDependencies
    });
  }
};
export default function CTF(
  ctfs: Array<CtfNode>,
  alfredConfig: AlfredConfig,
  interfaceState: InterfaceState
): CtfMap {
  const map: CtfMap = new Map();

  ctfs.forEach(ctfNode => {
    const ctfWithHelpers = {
      ...ctfNode,
      ...AddCtfHelpers
    };
    // eslint-disable-next-line
    ctfWithHelpers.interfaces = normalizeInterfacesOfSkill(ctfWithHelpers.interfaces);
    if (ctfWithHelpers.interfaces.length) {
      ctfWithHelpers.interfaces.forEach(e => {
        if (e.module.resolveSkill) {
          if (e.module.resolveSkill(ctfs, interfaceState) !== false) {
            map.set(ctfNode.name, ctfWithHelpers);
          }
        } else {
          map.set(ctfNode.name, ctfWithHelpers);
        }
      });
    } else {
      map.set(ctfNode.name, ctfWithHelpers);
    }
  });

  map.forEach(ctf => {
    Object.entries(ctf.ctfs || {}).forEach(([ctfName, ctfFn]) => {
      const correspondingCtfNode = map.get(ctfName);
      if (correspondingCtfNode) {
        map.set(
          ctfName,
          ctfFn(correspondingCtfNode, map, { alfredConfig, ...interfaceState })
        );
      }
    });
  });

  return map;
}

/*
 * Intended to be used for testing purposes
 */
export function getConfigs(ctf: CtfMap): Array<configType> {
  return Array.from(ctf.values())
    .map(ctfNode => ctfNode.configFiles || [])
    .reduce((p, c) => [...p, ...c], [])
    .map(e => e.config);
}

/**
 * Delete .configs dir
 */
export function deleteConfigs(): Promise<void> {
  if (fs.existsSync(configsBasePath)) {
    return new Promise(resolve => {
      rimraf(configsBasePath, () => {
        resolve();
      });
    });
  }
  return Promise.resolve();
}

/**
 * Write configs to a './.configs' directory
 */
export async function writeConfigsFromCtf(ctf: CtfMap): CtfMap {
  // Create a new .configs dir and write the configs
  const configs = Array.from(ctf.values())
    .map(ctfNode => ctfNode.configFiles || [])
    .reduce((p, c) => [...p, ...c], []);

  if (!fs.existsSync(configsBasePath)) {
    fs.mkdirSync(configsBasePath);
  }

  await Promise.all(
    configs
      .filter(config => config.write === true)
      .map(config => {
        const filePath = path.join(configsBasePath, config.path);
        const convertedConfig =
          typeof config.config === 'string'
            ? config.config
            : JSON.stringify(config.config);

        // Write sync to prevent data races when writing configs in parallel
        return fs.writeFileSync(filePath, convertedConfig);
      })
  );

  return ctf;
}

export function addSubCommandsToCtfNodes(ctf: CtfMap): CtfMap {
  ctf.forEach(ctfNode => {
    const subcommands = ctfNode.interfaces.map(e => require(e.name).subcommand); // eslint-disable-line
    ctfNode.subcommands = subcommands; // eslint-disable-line
  });
  return ctf;
}

/**
 * Intended to be used for testing purposes
 */
export function getDependencies(ctf: CtfMap): { [x: string]: string } {
  return Array.from(ctf.values())
    .map(ctfNode => ctfNode.dependencies || {})
    .reduce((p, c) => ({ ...p, ...c }), {});
}

export function getDevDependencies(ctf: CtfMap): { [x: string]: string } {
  return Array.from(ctf.values())
    .map(ctfNode => ctfNode.devDependencies || {})
    .reduce((p, c) => ({ ...p, ...c }), {});
}

export function execCommand(cmd: string) {
  return childProcess.execSync(cmd, { stdio: [0, 1, 2] });
}

export function getInterfaceForSubcommand(ctf: CtfMap, subcommand: string) {
  const interfaceForSubcommand = Array.from(ctf.values())
    .filter(
      ctfNode =>
        ctfNode.hooks && ctfNode.interfaces && ctfNode.interfaces.length
    )
    .reduce((arr, ctfNode) => arr.concat(ctfNode.interfaces.map(e => require(e.name))), []) // eslint-disable-line
    .find(ctfInterface => ctfInterface.subcommand === subcommand);

  if (!interfaceForSubcommand) {
    throw new Error(
      `The subcommand "${subcommand}" does not have an interface or the subcommand does not exist`
    );
  }

  return interfaceForSubcommand;
}

export function getExecuteWrittenConfigsMethods(
  ctf: CtfMap,
  interfaceState: InterfaceState
) {
  return Array.from(ctf.values())
    .filter(
      ctfNode =>
        ctfNode.hooks && ctfNode.interfaces && ctfNode.interfaces.length
    )
    .reduce((prev, curr) => prev.concat(curr), [])
    .map(ctfNode => {
      const configFiles = ctfNode.configFiles.map(configFile => ({
        ...configFile,
        path: path.join(configsBasePath, configFile.path)
      }));
      return ctfNode.interfaces.map(e => {
        const { subcommand } = require(e.name); // eslint-disable-line
        return {
          fn: (alfredConfig: AlfredConfig, flags: Array<string> = []) =>
            ctfNode.hooks.call({
              configFiles,
              ctf,
              alfredConfig,
              interfaceState,
              subcommand,
              flags
            }),
          // @HACK: If interfaces were defined, we could import the @alfredpkg/interface-*
          //        and use the `subcommand` property. This should be done after we have
          //        some interfaces to work with
          subcommand
        };
      });
    })
    .reduce((p, c) => p.concat(c), [])
    .reduce(
      (p, c) => ({
        ...p,
        [c.subcommand]: c.fn
      }),
      {}
    );
}
