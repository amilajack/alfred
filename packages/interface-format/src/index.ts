import { SkillInterfaceModule } from '@alfred/types';
import { interfaceResolvesSkillDefault } from '@alfred/helpers';

const skillInterface: SkillInterfaceModule = {
  subcommand: 'format',
  description: 'Format the source code of your app',
  resolveSkill: interfaceResolvesSkillDefault(
    'format',
    require('../package.json').name
  )
};

export default skillInterface;
