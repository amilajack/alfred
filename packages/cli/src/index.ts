import path from 'path';
import fs from 'fs';
import { formatPkgJson } from '@alfred/core';
import handlebars from 'handlebars';
import { Entrypoint, Target } from '@alfred/types';

const TEMPLATES_DIR = path.resolve(__dirname, '../templates');

export type ValidPkgNameResult = {
  validForNewPackages: boolean;
  validForOldPackages: boolean;
  errors?: string[];
  warnings?: string[];
};

export type RawTemplateData = {
  entrypoint?: Entrypoint;
  'alfred-pkg'?: {
    semver: string;
  };
};

export type TemplateData = {
  entrypoint: Entrypoint;
  'alfred-pkg': {
    semver: string;
  };
};

export type GitConfig = {
  user: {
    name: string;
    email: string;
  };
};

async function compileTemplate(
  templateFilename: string
): Promise<HandlebarsTemplateDelegate> {
  const source = await fs.promises.readFile(
    path.resolve(TEMPLATES_DIR, templateFilename)
  );
  return handlebars.compile(source.toString(), { noEscape: true });
}

export async function addEntrypoints(
  rawTemplateData: RawTemplateData,
  root: string,
  targets: Array<Target>
): Promise<void> {
  const [
    APP_TEMPLATE,
    LIB_TEMPLATE,
    APP_BROWSER_HTML_TEMPLATE,
    TEST_TEMPLATE
  ] = await Promise.all(
    ['app.js.hbs', 'lib.js.hbs', 'index.html.hbs', 'test.js.hbs'].map(
      compileTemplate
    )
  );

  await Promise.all(
    targets.map(target => {
      const { project, platform } = target;
      const isApp = project === 'app';
      const isBrowser = platform === 'browser';

      const templateData = {
        ...rawTemplateData,
        project: { project, target, isApp, isBrowser }
      };

      const writeIndexHtml = (): Promise<void> => {
        if (isApp && isBrowser) {
          const content = APP_BROWSER_HTML_TEMPLATE(templateData);
          return fs.promises.writeFile(
            path.join(root, './src/index.html'),
            content
          );
        }
        return Promise.resolve();
      };

      return Promise.all(
        [
          {
            file: `./src/${project}.${target}.js`,
            content: (isApp ? APP_TEMPLATE : LIB_TEMPLATE)(templateData)
          },
          {
            file: `./tests/${project}.${target}.spec.js`,
            content: TEST_TEMPLATE(templateData)
          }
        ]
          .map(({ file, content }) =>
            fs.promises.writeFile(path.join(root, file), content)
          )
          .concat([writeIndexHtml()])
      );
    })
  );
}

// This function lives here because it is used in both tests and the cli
export async function addBoilerplate(
  templateData: TemplateData,
  root: string
): Promise<void> {
  const [
    GITIGNORE_TEMPLATE,
    NPM_TEMPLATE,
    README_TEMPLATE,
    EDITORCONFIG_TEMPLATE
  ] = await Promise.all(
    [
      '.gitignore.hbs',
      'package.json.hbs',
      'README.md.hbs',
      '.editorconfig.hbs'
    ].map(compileTemplate)
  );

  await Promise.all(
    [
      {
        file: '.gitignore',
        content: GITIGNORE_TEMPLATE(templateData)
      },
      {
        file: '.editorconfig',
        content: EDITORCONFIG_TEMPLATE(templateData)
      },
      {
        file: 'README.md',
        content: README_TEMPLATE(templateData)
      }
    ].map(({ file, content }) =>
      fs.promises.writeFile(path.join(root, file), content)
    )
  );

  const pkgPath = path.join(root, 'package.json');
  const formattedPkg = await formatPkgJson(
    JSON.parse(NPM_TEMPLATE(templateData))
  );
  await fs.promises.writeFile(pkgPath, formattedPkg);
  const { project: project, platform: target } = templateData.entrypoint;

  await addEntrypoints(templateData, root, [
    { project: project, platform: target, env: 'production' }
  ]);
}

export function getSingleSubcommandFromArgs(args: Array<string>): string {
  if (args.length === 0) throw new Error('One subcommand must be passed');
  if (args.length > 1) throw new Error('Only one subcommand can be passed');
  return args[0];
}
