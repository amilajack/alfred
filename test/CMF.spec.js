// @flow
import * as CmfNodes from '../src/CMF';

const { getConfigs, default: CMF, getDependencies } = CmfNodes;

describe('CMF', () => {
  it('should do basic cmf', () => {
    const { babel, eslint, webpack } = CmfNodes;
    expect(getConfigs(CMF([babel, eslint, webpack]))).toMatchSnapshot();
    expect(getDependencies(CMF([babel, eslint, webpack]))).toMatchSnapshot();
  });
});
