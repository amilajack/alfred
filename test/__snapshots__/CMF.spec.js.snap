// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CMF should do basic cmf 1`] = `
Array [
  Object {
    "extends": "@babel/preset-env",
  },
  Object {
    "extends": Array [
      "bliss",
    ],
    "parser": "babel-eslint",
  },
  Object {
    "module": Object {
      "devtool": "source-map",
      "entry": "./app/main.dev",
      "mode": "production",
      "output": Object {
        "filename": "./app/main.prod.js",
        "path": "app",
      },
      "rules": Array [
        Object {
          "exclude": /node_modules/,
          "test": /\\\\\\.jsx\\?\\$/,
          "use": Object {
            "loader": "babel-loader",
            "options": Object {
              "cacheDirectory": true,
            },
          },
        },
      ],
      "target": "electron-main",
    },
  },
]
`;

exports[`CMF should do basic cmf 2`] = `
Object {
  "@babel/cli": "7.2.0",
  "@babel/core": "7.2.0",
  "@babel/preset": "env@7.2.0",
  "babel-eslint": "5.0.0",
  "babel-loader": "5.0.0",
  "eslint": "5.0.0",
  "webpack": "5.0.0",
}
`;
